articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,315 @@
+Given the non-negligible overhead incurred of multiple ``Participant``s per process, a change was introduced to map a single ``Participant`` to a context, and allow for multiple nodes to share that context.
+
+Based on the DDS Security specification v1.1, a ``Participant`` can only utilise a single security identity; consequently the access control permissions applicable to every node mapped to a given context must be consolidated and combined into a single set of security artifacts.
+As such, additional tooling and extensions to SROS 2 are necessary to support this new paradigm.
+
+## Keystore
+
+With the additional structure of contexts, it’s perhaps best to take the opportunity to restructure the keystore layout as well.
+Rather than a flat directory of namespaced node security directories, we can push all such security directories down into a designated `contexts` sub-folder.
+Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory.
+This is reminiscent of the pattern used earlier Keymint [1].
+
+
+```
+$ tree keystore/
+keystore
+├── contexts
+│   └── ...
+│       └── ...
+├── private
+│   ├── ca.csr.pem
+│   └── ca.key.pem
+└── public
+    ├── ca.cert.pem
+    ├── identity_ca.cert.pem -> ca.cert.pem
+    └── permissions_ca.cert.pem -> ca.cert.pem
+```
+
+
+### ``public``
+
+The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
+As such, this can be given read access to all executables.
+Note that in the default case, both the `identity_ca` and `permissions_ca` points to the same CA certificate.
+
+### ``private``
+
+The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
+This directory should be redacted before deploying the keystore onto the target device/robot.
+
+### ``contexts``
+
+The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
+Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
+The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.
+
+
+## Runtime
+
+Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
+To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 
+By convention, ros2launch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.
```

Comment:
Is this still something we consider ?
I thought we ended up moving away from any correlation between node name or namespace and context paths

Reply:
We moved away from node names. But the namespace in scope of the launch file could still be used to resolve relative context paths.

Reply:
I think that the same applies to namespaces.
If a context have several nodes, each of them created with a different `namespaces`, which of them will be used to expand the `security context`?

Reply:
> I think that the same applies to namespaces.

I think that was already addressed in that the context would be resolved at the point where the composing container is first instantiated in the launch file, e.g. what ever relative or absolute context the user prescribes when starting the container that may house nodes of various other namespaces. 

> > Nice, that's simple. So ComposableNode would just not have a context XML attribute?
>
> Exactly, it wouldn't be possible to have one as those components will be loaded in the same Context (a node container only has one Context where all the nodes are loaded).
https://github.com/ros2/design/pull/274/files#r385215843

From what you also mentioned prior, roslaunch does not yet have an API for instantiating containers, so that would also need to be addressed:

https://github.com/ros2/design/pull/274#discussion_r390290150

Reply:
Outside of the case of containers. I may want my nodes (standalone processes) to be pushed in a namespace without wanting them to look for security files in that same namespace. for example:
I want to have all my navigation nodes under the "navigation" namespace:
  
upstream launch file : 
awesome_navigation.launch.xml
```xml
<launch>
  <node ... awesome navigation node />
  <node static transform publisher context="tf_publisher" />
</launch>
```
my launchfile
```xml
<launch>
<group>
    <push_ros_namespace namespace="navigation"/>
    <include pkg="awesome_navigation" file="awesome_navigation.launch.xml"/>
<group>
</launch>
```

I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".
All I can do is:
- go modify the upstream launch file to use an absolute context
or
- create a `/foo/tf_publisher` set of artifacts to be used just by this node.

---

The story around composing nodes and launchfiles with security contexts is not clear, and neither the suggested approach or following namespaces or the alternatives allows to address it. It seems that the only discussed alternative allowing to modify included launchfiles contexts in a granular manner is to have to add a bunch of arguments to each launchfile

Reply:
> I may want my nodes (standalone processes) to be pushed in a namespace without wanting them to look for security files in that same namespace.

Some concrete reasons for doing so might help us polish our design ideas. 

> I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".

To be fair, as a user, you did choose to push the entire 3rd party launch file into a custom namespace, so I think it makes sense that the default behavior for all such context instances pushed would be to similarly push the relative context path they should look in.

> All I can do is

There is another alternative: to customize your keystore contexts structure to match the level of complexity that you'd like to achieve for the context lookup.
On could simply use links in the keystore:

```
cd ~/keystore/contexts/foo
ln -s ../tf_publisher tf_publisher
```

> It seems that the only discussed alternative allowing to modify included launchfiles contexts in a granular manner is to have to add a bunch of arguments to each launchfile

Which upstream package authors could choose or not choose to follow through with, that being I think most would skip on given how tediously verbose it quickly gets. (Security & Usability)

> I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".

I don't think it would be a good habit to have external launch files starting processes that use contexts orthogonal to namespace hierarchy, as it might get harder for the user to grok what part of the computational graph uses what security certificates.

Also given the static transform publisher would be remapped by the namespace scope (unless you pass a ns="/" arg that I didn't see), the node will need unique permissions like node param services that are FQN'ed.

In summery, if the launch file is in your control, then using absolute context paths could be a simple means of overriding name space scoping when complexity in context lookup is required. Alternatively, when the launch file is upstream, then you can still compensate by adjusting the keystore hierarchy which is still in your control.

Reply:
Relegated to future work: 3e148b2